File: src/main/java/com/jcloisterzone/game/capability/FamiliesCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.board.*;
import com.jcloisterzone.feature.City;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.feature.Structure;
import com.jcloisterzone.feature.modifier.StringStrictMergingModifier;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.setup.GameElementQuery;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import com.jcloisterzone.reducers.PlaceTile;

import io.vavr.collection.Stream;

import static com.jcloisterzone.XMLUtils.attributeIntValue;
import static com.jcloisterzone.XMLUtils.attributeStringValue;

import org.w3c.dom.Element;

public class FamiliesCapability extends Capability<Void> {

	public static final StringStrictMergingModifier FAMILY = new StringStrictMergingModifier("family", new GameElementQuery("families"));

	@Override
    public Feature initFeature(GameState state, String tileId, Feature feature, Element xml) {
        if (feature instanceof City) {
    	    if (attributeIntValue(xml, "pennants", 0) > 0 ) {
                String family = "blue";
                if (xml.hasAttribute("family")) {
                    family = xml.getAttribute("family");
                }
                feature = ((City) feature).putModifier(FAMILY, family);
    	    }
        }
        return feature;
    }

	@Override
    public boolean isTilePlacementAllowed(GameState state, Tile tile, PlacementOption placement) {
		
		if (!(FAMILY.getEnabledBy() != null && FAMILY.getEnabledBy().apply(state))) return true;
		     // Skip testing if FAMILY is not initalized by setup

		Position pos = placement.getPosition();
        Rotation rot = placement.getRotation();

        final GameState finalState = (new PlaceTile(tile, pos, rot)).apply(state);

        Stream<City> cities = finalState.getTileFeatures2(pos, Structure.class)
        	    .filter(fp -> fp._2 instanceof City && ((City) fp._2).getModifier(finalState, City.PENNANTS, 0) > 0)
        	    .map(fp -> (City) fp._2)
        	    .toStream();
        
        if (cities.isEmpty()) {
        	// No city with Pennant, placement allowed
        	return true;
        }
        
        return !cities.exists(city ->
            city.getModifier(finalState, FAMILY, null).equals(StringStrictMergingModifier.CONFLICT)
        );
    }
}

File: src/main/ts/com/jcloisterzone/game/capability/FamiliesCapability.ts
import { attributeIntValue, type XmlElement } from "../../XmlUtils.js";
import type { PlacementOption } from "../../board/PlacementOption.js";
import type { Tile } from "../../board/Tile.js";
import { City } from "../../feature/City.js";
import type { Feature } from "../../feature/Feature.js";
import { FamiliesModifier } from "../../feature/modifier/FamiliesModifier.js";
import { isInstanceOfStructure } from "../../feature/Structure.js";
import { PlaceTile } from "../../reducers/PlaceTile.js";
import { GameElementQuery } from "../setup/GameElementQuery.js";
import { Capability } from "../Capability.js";
import type { GameState } from "../state/GameState.js";

/** Family Feud: a city with pennant(s) carries a family colour (FAMILY modifier); a tile
 *  placement that would merge cities of two different (non-grey) families is forbidden. */
export class FamiliesCapability extends Capability<void> {
  static readonly FAMILY = new FamiliesModifier("family", new GameElementQuery("families"));

  override initFeature(_state: GameState, _tileId: string, feature: Feature, xml: XmlElement): Feature {
    if (feature instanceof City) {
      if ((attributeIntValue(xml, "pennants", 0) ?? 0) > 0) {
        const family = xml.hasAttribute("family") ? xml.getAttribute("family") : "blue";
        feature = feature.putModifier(FamiliesCapability.FAMILY, family);
      }
    }
    return feature;
  }

  override isTilePlacementAllowed(state: GameState, tile: Tile, placement: PlacementOption): boolean {
    const enabledBy = FamiliesCapability.FAMILY.getEnabledBy();
    // skip testing if FAMILY is not initialised by setup
    if (!(enabledBy !== null && enabledBy.apply(state))) return true;

    const pos = placement.getPosition();
    const rot = placement.getRotation();
    const finalState = new PlaceTile(tile, pos, rot).apply(state);

    const cities: City[] = [];
    for (const t of finalState.getTileFeatures2(pos)) {
      const f = t._2;
      if (!isInstanceOfStructure(f)) continue;
      if (
        f instanceof City &&
        !f.hasModifier(finalState, City.ELIMINATED_PENNANTS) &&
        f.getModifier(finalState, City.PENNANTS, 0) > 0
      ) {
        cities.push(f);
      }
    }

    if (cities.length === 0) return true; // no city with a pennant → allowed

    return !cities.some(
      (city) => city.getModifier(finalState, FamiliesCapability.FAMILY, null) === FamiliesModifier.CONFLICT,
    );
  }
}

Capability.register(FamiliesCapability);
